include/modem/sms.h

```diff
@@ -24,52 +24,168 @@ extern "C" {
+ */
+struct sms_time {
+	uint8_t year;    /** @brief Year. Last two digits of the year.*/
+	uint8_t month;   /** @brief Month. */
+	uint8_t day;     /** @brief Day. */
+	uint8_t hour;    /** @brief Hour. */
+	uint8_t minute;  /** @brief Minute. */
+	uint8_t second;  /** @brief Second. */
+	int8_t timezone; /** @brief Timezone. */
+};
+
+/**
+ * @brief SMS address, i.e., phone number.
+ * 
+ * @details This may represent either originating or destination address and is
+ * specified in 3GPP TS 23.040 Section 9.1.2.5.
+ */
+struct sms_address {
+	/** @brief Address in NUL-terminated string format. */
+	char    address_str[SMS_MAX_ADDRESS_LEN_CHARS + 1];
+	/**
+	 * @brief Address in semi-octet representation specified in
+	 * 3GPP TS 23.040 Section 9.1.2.3.
+	 * 
+	 * TODO: Just remove this field?
+	 */
+	uint8_t address[SMS_MAX_ADDRESS_LEN_OCTETS];
+	/** @brief Address length in number of characters. */
+	uint8_t length;
+	/** @brief Address type as specified in 3GPP TS 23.040 Section 9.1.2.5. */
+	uint8_t type;
+};
+
+/**
+ * @brief SMS concatenated short message information.
+ * 
+ * @details This is specified in 3GPP TS 23.040 Section 9.2.3.24.1 and 9.2.3.24.8.
+ */
+struct sms_udh_concatenated {
+	/** @brief Indicates whether this field is present in the SMS message. */
+	bool present;
+	/** @brief Concatenated short message reference number. */
+	uint16_t ref_number;
+	/** @brief Maximum number of short messages in the concatenated short message. */
+	uint8_t total_msgs;
+	/** @brief Sequence number of the current short message. */
+	uint8_t seq_number;
+};
+
+/**
+ * @brief SMS application port addressing information.
+ * 
+ * @details This is specified in 3GPP TS 23.040 Section 9.2.3.24.3 and 9.2.3.24.4.
+ */
+struct sms_udh_app_port {
+	/** @brief Indicates whether this field is present in the SMS message. */
+	bool present;
+	/** @brief Destination port. */
+	uint16_t dest_port;
+	/** @brief Source port. */
+	uint16_t src_port;
+};
+
+/**
+ * SMS-DELIVER message header.
+ * This is for incoming SMS message and more specifically SMS-DELIVER
+ * message specified in 3GPP TS 23.040.
+ */
+struct sms_deliver_header {
+	/** @brief Timestamp. */
+	struct sms_time time;
+	/** @brief Originating address, i.e., phone number. */
+	struct sms_address originating_address;
+	/** @brief Application port addressing information. */
+	struct sms_udh_app_port app_port;
+	/** @brief Concatenated short message information. */
+	struct sms_udh_concatenated concatenated;
+};
+
+/**
+ * @brief SMS header.
+ * 
+ * @details This can easily be extended to support additional message types.
+ */
+union sms_header {
+	struct sms_deliver_header deliver;
+};
+
 /** @brief SMS PDU data. */
 struct sms_data {
-	char *alpha;
-	uint16_t length;
-	char *pdu;
+	/** Received message type. */
+	enum sms_type type;
+	/** SMS header. */
+	union sms_header header;
+
+	/** Length of the data in data buffer. */
+	int data_len;
```

Comment:
Maybe rename this to payload_len and payload? The word "data" is a bit indistinct, because you could also say that the header data.

Reply:
Likely using payload

Reply:
I'm using payload here now